core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1470,21 +1503,45 @@ class KafkaApis(val requestChannel: RequestChannel,
     val transactionalId = addPartitionsToTxnRequest.transactionalId
     val partitionsToAdd = addPartitionsToTxnRequest.partitions
 
-    // Send response callback
-    def sendResponseCallback(error: Errors): Unit = {
-      def createResponse(throttleTimeMs: Int): AbstractResponse = {
-        val responseBody: AddPartitionsToTxnResponse = new AddPartitionsToTxnResponse(throttleTimeMs, error)
-        trace(s"Completed $transactionalId's AddPartitionsToTxnRequest with partitions $partitionsToAdd: errors: $error from client ${request.header.clientId}")
-        responseBody
+    if(!authorize(request.session, Write, new Resource(ProducerTransactionalId, transactionalId)))
+      sendResponseMaybeThrottle(request, (throttleTimeMs: Int) => new InitPidResponse(throttleTimeMs, Errors.TRANSACTIONAL_ID_AUTHORIZATION_FAILED))
+    else {
+      val internalTopics = partitionsToAdd.asScala.filter {tp => kafka.common.Topic.InternalTopics.contains(tp.topic())}
+
+      val (existingAndAuthorizedForDescribeTopics, nonExistingOrUnauthorizedForDescribeTopics) =
```

Comment:
For the regular offset commit API, we just repeat the same error for all topic partitions. I'd probably just do the same thing with this API so that they stay relatively close.